common/common_test.go

```diff
@@ -0,0 +1,290 @@
+			testT.Error("unexpected sleeps", sleepCount, sleepDuration)
+		}
+		// should re-request the same height
+		if height != "22" {
+			testT.Error("unexpected height")
+		}
+		// Back to sunny-day
+		return blocks[2], nil
+	case 5:
+		if height != "23" {
+			testT.Error("unexpected height")
+		}
+		// Simulate a reorg (it doesn't matter which block we return here, as
+		// long as its prevhash doesn't match the latest block's hash)
+		return blocks[2], nil
+	case 6:
+		// When a reorg occurs, the ingestor backs up 2 blocks
+		if height != "21" { // 23 - 2
+			testT.Error("unexpected height")
+		}
+		return blocks[1], nil
+	case 7:
+		if height != "22" {
+			testT.Error("unexpected height")
+		}
+		// Should fail to Unmarshal the block, sleep, retry
+		return nil, nil
+	case 8:
+		if sleepCount != 2 || sleepDuration != 20*time.Second {
+			testT.Error("unexpected sleeps", sleepCount, sleepDuration)
+		}
+		if height != "22" {
+			testT.Error("unexpected height")
+		}
+		// Back to sunny-day
+		return blocks[2], nil
+	}
+	if height != "23" {
+		testT.Error("unexpected height")
+	}
+	testT.Error("getblockStub called too many times")
+	return nil, nil
+}
+
+func TestBlockIngestor(t *testing.T) {
+	testT = t
+	RawRequest = getblockStub
+	Sleep = sleepStub
+	testcache := NewBlockCache(4)
+	BlockIngestor(testcache, 20, 7)
+	if step != 7 {
+		t.Error("unexpected final step", step)
+	}
+	step = 0
+	sleepCount = 0
+	sleepDuration = 0
+}
+
+func TestGetBlockRange(t *testing.T) {
+	testT = t
+	RawRequest = getblockStub
+	testcache := NewBlockCache(4)
+	blockChan := make(chan walletrpc.CompactBlock)
+	errChan := make(chan error)
+	go GetBlockRange(testcache, blockChan, errChan, 20, 22)
+
+	// read in block 20
+	select {
+	case err := <-errChan:
+		// this will also catch context.DeadlineExceeded from the timeout
+		t.Fatal("unexpected error:", err)
+	case cBlock := <-blockChan:
+		if cBlock.Height != 380640 {
+			t.Fatal("unexpected Height:", cBlock.Height)
+		}
+	}
+
+	// read in block 21
+	select {
+	case err := <-errChan:
+		// this will also catch context.DeadlineExceeded from the timeout
+		t.Fatal("unexpected error:", err)
+	case cBlock := <-blockChan:
+		if cBlock.Height != 380641 {
+			t.Fatal("unexpected Height:", cBlock.Height)
+		}
+	}
+
+	// try to read in block 22, but this will fail (see case 3 above)
+	select {
+	case err := <-errChan:
+		// this will also catch context.DeadlineExceeded from the timeout
+		if err.Error() != "block requested is newer than latest block" {
+			t.Fatal("unexpected error:", err)
+		}
+	case _ = <-blockChan:
+		t.Fatal("reading height 22 should have failed")
+	}
+
+	// check goroutine GetBlockRange() reaching the end of the range (and exiting)
+	go GetBlockRange(testcache, blockChan, errChan, 1, 0)
+	err := <-errChan
+	if err != nil {
+		t.Fatal("unexpected err return")
+	}
+}
+
+func TestGenerateCerts(t *testing.T) {
+	if GenerateCerts() == nil {
+		t.Fatal("GenerateCerts returned nil")
+	}
```

Comment:
I don't know a way to verify that the correct things happened, but at least we're running through the code.

Reply:
I opened https://github.com/zcash/lightwalletd/issues/181 so that we're less likely to mistakenly think GenerateCerts is fully tested when we look at the coverage (it'd be obvious if this code was broken because the certs wouldn't work so this is fine for now).